perses/rjmc/topology_proposal.py

```diff
@@ -910,6 +910,10 @@ def _identify_differences(self, topology, residue_map):
             # template_atom_names : dict, key : template atom index, value : template atom name
             template_atom_names = {}
             for atom in template.atoms:
+                if replace_with == 'LYN' and atom.name == 'HZ3': # Rename HZ3 to HZ1 s.t. the topology matches the naming in amino_acid_templates/LYN.pdb
+                    atom.name = 'HZ1'
+                    template.atomIndices['HZ1'] = template.atomIndices['HZ3']
+                    del template.atomIndices['HZ3']
```

Comment:
Will this ever find more than one match? Deleting things as you iterate can cause issues, it might be better to first loop over the atoms to rename them.
I don't understand why `template.atomIndices['HZ1'] = template.atomIndices['HZ3']` is in the loop since it operates on `template.atomIndices` and not `atom`, plus then if you find another atom with that name then you will get a key error.
Does this work?

```suggestion
                if replace_with == 'LYN' and atom.name == 'HZ3': # Rename HZ3 to HZ1 s.t. the topology matches the naming in amino_acid_templates/LYN.pdb
                    atom.name = 'HZ1'
            template.atomIndices['HZ1'] = template.atomIndices['HZ3']
            del template.atomIndices['HZ3']
```

Reply:
Or if the logic of this only needs to happen then `replace_with == 'LYN'` might have to be moved

Reply:
This logic should only happen for `LYN`. Basically the problem here is that the `LYN` amber template atoms do match the atoms in `LYN.pdb` (generated by pymol). Since LYN is deprotonated LYS, this means that terminal carbon on the LYN side chain should only contain 2 hydrogens, not 3. In the amber template, `LYN` contains HZ2 and HZ3, whereas in `LYN.pdb`, it contains HZ1 and HZ2. So here I was trying to rename the HZ3 atom (there's only one of them) in the amber template for LYN. I also need to update the atom indices dictionary (key: atom name, value: atom index), because otherwise the dictionary will contain the wrong atom name. I think my original implementation works better than your suggestion, because your suggestion allows the `template.atomIndices` dict to be changed for the HZ1 atom in any residue (which is not the behavior we want). Hopefully this clarifies things, but let me know if you still think some of the logic can be improved upon

Reply:
How does having `template.atomIndices['HZ1'] = template.atomIndices['HZ3']` in the loop ensure that HZ1 atom is changed only for the LYN residue? Since we are only looping over `for atom in template.atoms` I would expect that it would be okay if this was outside of the loop.

Okay I looked at more context and see now that the check  `replace_with == 'LYN'` is what is guarding which residue we are changing. 

@zhang-ivy  would you mind reorganizing this slightly so it is more clear when we are doing things for only one residue? something like this (I can't suggest changes that don't happen in the PR which is why I can't use that feature here).

``` python
for k, (residue, replace_with) in enumerate(residue_map):
            # Load residue template for residue to replace with
            if replace_with =='HIS':
                replace_with = 'HIE'
            template = self._templates[replace_with]
            # template_atom_names : dict, key : template atom index, value : template atom name
            template_atom_names = {}
            if replace_with == 'LYN':
                for atom in template.atoms:
                    if atom.name == 'HZ3': # Rename HZ3 to HZ1 s.t. the topology matches the naming in amino_acid_templates/LYN.pdb
                        atom.name = 'HZ1'
                 template.atomIndices['HZ1'] = template.atomIndices['HZ3']
                 del template.atomIndices['HZ3']
```
I think something like that would work? Moving the `if` logic to be as high as possible to modify things per-residue makes it more clear what is going on.

Reply:
The reason why I didnt implement it like your suggestion above is because I was trying to avoid adding an extra for loop that iterates over the template.atoms, since [there is already a for loop](https://github.com/choderalab/perses/blob/master/perses/rjmc/topology_proposal.py#L932-L933) that iterates over the atoms. 
If you feel strongly that your suggestion is more clear, we I can replace my original implementation with yours, however I think your implementation is a bit redundant.

Reply:
Or we could keep my implementation and I can add better comments that clarify what is happening the code? 
Something like `iterate over atoms in the mutant residue template. if mutant is LYN, we need to edit one of the atom names"

Reply:
Mike and I moved this discussion. We settled on the changes in the following commit.